test/lib/base: Reduce redundant build steps

iwaseyusuke opened the pull request:
Currently, the build steps of "go get" and removing the downloaded directory are not effective steps.
For example, "go get" for downloading dependencies can be replaced with "dep ensure".

Signed-off-by: IWASE Yusuke <iwase.yusuke0@gmail.com>

fujita commented:
IIRC, dep uses libraries at $GOPATH when possible (otherwise download remotely). When you execute this script locally, a docker image having libraries in cache makes image building faster?

iwaseyusuke commented:
Yes, the caches are some effective, but "dep ensure" does not seem to use those caches as expected.
Each time most of packages seem to be downloaded with "dep ensure".

The result on my environment: "dep ensure" seems to be fast enough without using caches.

|            | master    | patched   |
| ---------- | --------- | --------- |
| no cache   | 6m1.879s  | 3m42.299s |
| with cache | 3m16.645s | 3m31.700s |

Logs on master branch:

```
$ time fab -f ./test/lib/base.py make_gobgp_ctn
[localhost] local: echo 'FROM osrg/quagga
RUN go get -d github.com/osrg/gobgp/...; exit 0
RUN rm -rf /go/src/github.com/osrg/gobgp
RUN go get -u github.com/golang/dep/cmd/dep
ADD gobgp /go/src/github.com/osrg/gobgp/
RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp' > Dockerfile
[localhost] local: docker build -t gobgp .
Sending build context to Docker daemon  450.1MB
Step 1/6 : FROM osrg/quagga
 ---> a1fa42e1939e
Step 2/6 : RUN go get -d github.com/osrg/gobgp/...; exit 0
 ---> Running in 7f015cde5b0a
 ---> eec3a2b17bbc
Removing intermediate container 7f015cde5b0a
Step 3/6 : RUN rm -rf /go/src/github.com/osrg/gobgp
 ---> Running in 3b37b5489c88
 ---> 03e04dbc04b4
Removing intermediate container 3b37b5489c88
Step 4/6 : RUN go get -u github.com/golang/dep/cmd/dep
 ---> Running in 5c1b381d562b
 ---> 2bdba6a945a6
Removing intermediate container 5c1b381d562b
Step 5/6 : ADD gobgp /go/src/github.com/osrg/gobgp/
 ---> 9accac7398a9
Removing intermediate container 0f8d1e6e17f5
Step 6/6 : RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp
 ---> Running in 6d194a0f6eac
 ---> c8b4f7186c3b
Removing intermediate container 6d194a0f6eac
Successfully built c8b4f7186c3b
Successfully tagged gobgp:latest
[localhost] local: rm Dockerfile

Done.

real	6m1.879s
user	0m1.204s
sys	0m0.752s
$ 
$ 
$ git stash pop
On branch master
Your branch is up-to-date with 'upsteam/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   server/server.go

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (5782038d4eb4afa5171754fca8b2e64dcf3ed475)
$ 
$ 
$ time fab -f ./test/lib/base.py make_gobgp_ctn
[localhost] local: echo 'FROM osrg/quagga
RUN go get -d github.com/osrg/gobgp/...; exit 0
RUN rm -rf /go/src/github.com/osrg/gobgp
RUN go get -u github.com/golang/dep/cmd/dep
ADD gobgp /go/src/github.com/osrg/gobgp/
RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp' > Dockerfile
[localhost] local: docker build -t gobgp .
Sending build context to Docker daemon  450.1MB
Step 1/6 : FROM osrg/quagga
 ---> a1fa42e1939e
Step 2/6 : RUN go get -d github.com/osrg/gobgp/...; exit 0
 ---> Using cache
 ---> eec3a2b17bbc
Step 3/6 : RUN rm -rf /go/src/github.com/osrg/gobgp
 ---> Using cache
 ---> 03e04dbc04b4
Step 4/6 : RUN go get -u github.com/golang/dep/cmd/dep
 ---> Using cache
 ---> 2bdba6a945a6
Step 5/6 : ADD gobgp /go/src/github.com/osrg/gobgp/
 ---> 3598e02e1abd
Removing intermediate container 1975ffdbcf47
Step 6/6 : RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp
 ---> Running in 90751dd33591
 ---> e7706c6cd476
Removing intermediate container 90751dd33591
Successfully built e7706c6cd476
Successfully tagged gobgp:latest
[localhost] local: rm Dockerfile

Done.

real	3m16.645s
user	0m1.152s
sys	0m0.676s
```

Logs on patched branch:

```
$ time fab -f ./test/lib/base.py make_gobgp_ctn
[localhost] local: echo 'FROM osrg/quagga
RUN go get -u github.com/golang/dep/cmd/dep
RUN mkdir -p /go/src/github.com/osrg/
ADD gobgp /go/src/github.com/osrg/gobgp/
RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp' > Dockerfile
[localhost] local: docker build -t gobgp .
Sending build context to Docker daemon  450.1MB
Step 1/5 : FROM osrg/quagga
 ---> a1fa42e1939e
Step 2/5 : RUN go get -u github.com/golang/dep/cmd/dep
 ---> Running in 052b7b2087e2
 ---> 7206b2cd47c8
Removing intermediate container 052b7b2087e2
Step 3/5 : RUN mkdir -p /go/src/github.com/osrg/
 ---> Running in 61f4c1457d14
 ---> 1d4393c0d4ce
Removing intermediate container 61f4c1457d14
Step 4/5 : ADD gobgp /go/src/github.com/osrg/gobgp/
 ---> 8bf0526359c9
Removing intermediate container 41aa6c0f3268
Step 5/5 : RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp
 ---> Running in db50fd9ab55f
 ---> 7a054738c74d
Removing intermediate container db50fd9ab55f
Successfully built 7a054738c74d
Successfully tagged gobgp:latest
[localhost] local: rm Dockerfile

Done.

real	3m42.299s
user	0m1.220s
sys	0m0.832s
$ 
$ 
$ git stash pop
On branch test-lib-base-Reduce_redundant_build_steps
Your branch is up-to-date with 'origin/test-lib-base-Reduce_redundant_build_steps'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   server/server.go

no changes added to commit (use "git add" and/or "git commit -a")
Dropped refs/stash@{0} (e5667a3c912f72a1286524d1aa2fa2e940cdda5f)
$ 
$ 
$ time fab -f ./test/lib/base.py make_gobgp_ctn
[localhost] local: echo 'FROM osrg/quagga
RUN go get -u github.com/golang/dep/cmd/dep
RUN mkdir -p /go/src/github.com/osrg/
ADD gobgp /go/src/github.com/osrg/gobgp/
RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp' > Dockerfile
[localhost] local: docker build -t gobgp .
Sending build context to Docker daemon  450.1MB
Step 1/5 : FROM osrg/quagga
 ---> a1fa42e1939e
Step 2/5 : RUN go get -u github.com/golang/dep/cmd/dep
 ---> Using cache
 ---> 7206b2cd47c8
Step 3/5 : RUN mkdir -p /go/src/github.com/osrg/
 ---> Using cache
 ---> 1d4393c0d4ce
Step 4/5 : ADD gobgp /go/src/github.com/osrg/gobgp/
 ---> bf1ad6186c8a
Removing intermediate container 3d075e1bc560
Step 5/5 : RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp
 ---> Running in 84bb6780ee04
 ---> 6b0d7544e63c
Removing intermediate container 84bb6780ee04
Successfully built 6b0d7544e63c
Successfully tagged gobgp:latest
[localhost] local: rm Dockerfile

Done.

real	3m31.700s
user	0m1.224s
sys	0m0.616s
```

fujita commented:
Hmm, you meant that `dep ensure` always accesses remotely?
Then, after `go get`, how about moving the libraries to vendor directory?

iwaseyusuke commented:
Yes... I guess so.

With moving the downloaded packages into "vendor" directory, it took longer than before...

```
$ time fab -f ./test/lib/base.py make_gobgp_ctn
[localhost] local: echo 'FROM osrg/quagga
RUN go get -d github.com/osrg/gobgp/...; exit 0
RUN rm -rf /go/src/github.com/osrg/gobgp
RUN go get -u github.com/golang/dep/cmd/dep
ADD gobgp /go/src/github.com/osrg/gobgp/
RUN mkdir -p /go/src/github.com/osrg/gobgp/vendor/github.com
RUN ls -d /go/src/* | grep -v github.com | xargs -I{} mv {} /go/src/github.com/osrg/gobgp/vendor
RUN ls -d /go/src/github.com/* | grep -v osrg | xargs -I{} mv {} /go/src/github.com/osrg/gobgp/vendor/github.com
RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp' > Dockerfile
[localhost] local: docker build -t gobgp .
Sending build context to Docker daemon  450.7MB
Step 1/9 : FROM osrg/quagga
 ---> a1fa42e1939e
Step 2/9 : RUN go get -d github.com/osrg/gobgp/...; exit 0
 ---> Using cache
 ---> dc76e4f2dc4b
Step 3/9 : RUN rm -rf /go/src/github.com/osrg/gobgp
 ---> Using cache
 ---> 772c2c96c5fd
Step 4/9 : RUN go get -u github.com/golang/dep/cmd/dep
 ---> Using cache
 ---> 58ca90086275
Step 5/9 : ADD gobgp /go/src/github.com/osrg/gobgp/
 ---> 2b1cb4c14f4f
Removing intermediate container 23d043dd140e
Step 6/9 : RUN mkdir -p /go/src/github.com/osrg/gobgp/vendor/github.com
 ---> Running in 5394d95ad245
 ---> 40665bf607d7
Removing intermediate container 5394d95ad245
Step 7/9 : RUN ls -d /go/src/* | grep -v github.com | xargs -I{} mv {} /go/src/github.com/osrg/gobgp/vendor
 ---> Running in 514d440e17cc
 ---> 837b75a8babc
Removing intermediate container 514d440e17cc
Step 8/9 : RUN ls -d /go/src/github.com/* | grep -v osrg | xargs -I{} mv {} /go/src/github.com/osrg/gobgp/vendor/github.com
 ---> Running in 077a6125ae0e
 ---> 3ace016ee46d
Removing intermediate container 077a6125ae0e
Step 9/9 : RUN cd /go/src/github.com/osrg/gobgp && dep ensure && go install ./gobgpd ./gobgp
 ---> Running in 7779acd6d8f2
 ---> 9887c77bac01
Removing intermediate container 7779acd6d8f2
Successfully built 9887c77bac01
Successfully tagged gobgp:latest
[localhost] local: rm Dockerfile

Done.

real	8m20.348s
user	0m1.436s
sys	0m1.244s
```

fujita commented:
Pushed because of: https://github.com/golang/dep/blob/master/docs/FAQ.md#how-do-i-make-dep-resolve-dependencies-from-my-gopath

Seems that sorta caching under development so we could do something for better performance:
https://github.com/golang/dep/blob/master/docs/FAQ.md#why-is-dep-slow

iwaseyusuke commented:
Thank you for your instruction.
I understood that as far as the initial build like when building Docker images, the cache of `go get` cannot accelerate `dep` command enough.
